src/WebJobs.Extensions.DurableTask/Listener/FunctionExecutionHelper.cs

```diff
@@ -24,42 +25,43 @@ internal static class FunctionExecutionHelper
 
             try
             {
-                bool executedUserCode = false;
-                var triggeredFunctionData = new TriggeredFunctionData()
-                {
-                    TriggerValue = triggerInput.TriggerValue,
-                    ParentId = triggerInput.ParentId,
-                    InvokeHandler = userCodeHandler =>
-                    {
-                        executedUserCode = true;
-                        return triggerInput.InvokeHandler(userCodeHandler);
-                    },
-#if !FUNCTIONS_V1
-                    TriggerDetails = triggerInput.TriggerDetails,
-#endif
-                };
-#pragma warning restore CS0618
+                context.ExecutorCalledBack = false;
 
                 FunctionResult result = await executor.TryExecuteAsync(triggerInput, cancellationToken);
 
-                if (result.Succeeded)
+                if (context.ExecutorCalledBack)
                 {
-                    return WrappedFunctionResult.Success();
-                }
+                    if (result.Succeeded)
+                    {
+                        return WrappedFunctionResult.Success();
+                    }
 
-                if (cancellationToken.IsCancellationRequested)
-                {
-                    return WrappedFunctionResult.FunctionRuntimeFailure(result.Exception);
-                }
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        return WrappedFunctionResult.FunctionRuntimeFailure(result.Exception);
+                    }
 
-                if (!executedUserCode)
-                {
-                    WrappedFunctionResult.UserCodeFailure(new InvalidOperationException(
-                         "The function failed to start executing. " +
-                         "For .NET functions, this can happen if an unhandled exception is thrown in the function's class constructor."));
+                    return WrappedFunctionResult.UserCodeFailure(result.Exception);
                 }
+                else
+                {
+                    // This can happen if the constructor for a non-static function fails.
+                    // We want to treat this case exactly as if the function itself is throwing the exception.
+                    // So we execute the middleware directly, instead of via the executor.
+                    try
+                    {
+                        var exception = result.Exception ?? new InvalidOperationException("The function failed to start executing.");
 
-                return WrappedFunctionResult.UserCodeFailure(result.Exception);
+                        await triggerInput.InvokeHandler(() => Task.FromException<object>(exception));
+
+                        return WrappedFunctionResult.Success();
```

Comment:
So the idea is that some entity functionality will not actually throw an exception even when we manually invoke the handled with a thrown exception. In those cases, we want to treat the operation as a success?

Reply:
Right. For entities there is not a one-to-one correspondence between calling the middleware executor (which corresponds to delivering a message to the entity scheduler) and the function execution (which corresponds to executing an operation on the entity). Some messages do not cause an entity operation execution, and some messages can cause more than one entity operation execution. And yes, the entity operation exceptions are caught by the entity scheduler.

Reply:
An interesting effect of this is that you can use an undeclared entity class as a lock.